core/src/main/java/com/scalar/db/transaction/consensuscommit/CommitHandler.java

```diff
@@ -41,38 +43,37 @@ public CommitHandler(
   }
 
   public void commit(Snapshot snapshot) throws CommitException, UnknownTransactionStatusException {
-    prepare(snapshot, true);
+    try {
+      prepare(snapshot);
+    } catch (PreparationException e) {
+      abort(snapshot.getId());
+      rollbackRecords(snapshot);
+      if (e instanceof PreparationConflictException) {
+        throw new CommitConflictException(e.getMessage(), e);
+      }
+      throw new CommitException(e.getMessage(), e);
+    }
+
     preCommitValidation(snapshot, true);
     commitState(snapshot);
     commitRecords(snapshot);
   }
 
-  public void prepare(Snapshot snapshot, boolean abortIfError)
-      throws CommitException, UnknownTransactionStatusException {
-    String id = snapshot.getId();
+  public void prepare(Snapshot snapshot) throws PreparationException {
     try {
       prepareRecords(snapshot);
-    } catch (Exception e) {
+    } catch (NoMutationException e) {
+      throw new PreparationConflictException("preparing record exists", e);
+    } catch (RetriableExecutionException e) {
+      throw new PreparationConflictException("conflict happened when preparing records", e);
+    } catch (ExecutionException e) {
       logger.warn("preparing records failed", e);
```

Comment:
This warning won't be logged in the new source code if the exception isn't ExecutionException. Is it okay?

BTW, the cause exception is contained in `PreparationException`, so I'm not sure this warning is needed.

Reply:
Agree. I will remove this log. Thanks.

Reply:
Deleted the log in 77c422608ec17743dbd74fd3a3f858df71ca40c0. Thanks.